src/libsyntax/parse/parser.rs

```diff
@@ -4726,9 +4726,13 @@ impl<'a> Parser<'a> {
             let fields = try!(self.parse_record_struct_body(&class_name));
             (fields, None)
         // Tuple-style struct definition with optional where-clause.
-        } else {
+        } else if self.token == token::OpenDelim(token::Paren) {
             let fields = try!(self.parse_tuple_struct_body(&class_name, &mut generics));
             (fields, Some(ast::DUMMY_NODE_ID))
+        } else {
+            let token_str = self.this_token_to_string();
+            return Err(self.fatal(&format!("expected `where`, `{}`, `(`, or `;` after struct \
+                                            name, found `{}`", "{", token_str)))
```

Comment:
I'd say I can fix { escaping in this PR, in two places (there is a second similar error message in `parse_record_struct_body`).

 As for the removing the explicit error check, I'm not so sure:
* this PR does not change behavior at all, it just rearenges/removes code. Changes to code and test are better done in separate PR, if needed. 
* this eror message looks better then the autogenerated one.